[TAN-4989] Whitelist selected image metadata (to avoid messing up colours)

jinjagit opened the pull request:
See technical discussion, [Permit ICC color profile?](https://www.notion.so/govocal/Permit-image-ICC-color-profile-2329663b7b2680188193f2ea3724fec3), and ticket convo, for explanation of choice of approach.

- Now maintaining vibrancy (though images displayed in app are still lower resolution than originals, ofc)
- Also permits Orientation EXIF tag, removing need for rotation of images before storing.

- Includes Sentry error reporting, that gives errors like [this](https://sentry.hq.citizenlab.co/organizations/citizenlab/issues/101585/?environment=development&project=2&query=is%3Aunresolved+is%3Afor_review+assigned_or_suggested%3A%5Bme%2C+none%5D&sort=inbox), including additional data, such as:
```
exiftool_command | exiftool -all= -tagsFromFile @ -icc_profile -orientation -overwrite_original blah /cl2_back/tmp/1752833049-253774638011973-0002-1866/image.jpeg
-- | --
exiftool_error_signal | None
exiftool_exit_status | 1
exiftool_stderr | Warning: Error opening file - blah Error: File not found - blah Warning: ICC_Profile deleted. Image colors may be affected - /cl2_back/tmp/1752833049-253774638011973-0002-1866/image.jpeg
exiftool_stdout | 1 image files updated     1 files weren't updated due to errors
file_path | /cl2_back/tmp/1752833049-253774638011973-0002-1866/image.jpeg
```
- Will also raise a 2nd Runtime error like [this](https://sentry.hq.citizenlab.co/organizations/citizenlab/issues/101586/?environment=development&project=2&query=is%3Aunresolved), which also gets reported in Sentry. This is because without the raising of this error, no error was seen in the browser request to upload the file, and neither Sentry error was logged (maybe the latter is a dev env thing?)

Before this change (uploaded & downsized displayed image on left, original on right):
<img width="553" height="215" alt="Screenshot 2025-07-16 at 16 05 56" src="https://github.com/user-attachments/assets/8f0b9107-8032-43a5-9375-c63360b3a513" />

After this change (uploaded & downsized displayed image on left, original on right):
<img width="536" height="215" alt="Screenshot 2025-07-16 at 18 31 00" src="https://github.com/user-attachments/assets/b160e5c9-fcd7-485b-8310-237aba8dd2a1" />

After this change - uploaded image without downsizing:
<img width="363" height="311" alt="Screenshot 2025-07-17 at 10 28 00" src="https://github.com/user-attachments/assets/57b1e2f3-eeed-4a71-a862-99764741c289" />

# Changelog
## Fixed
- [TAN-4989] Whitelist selected image metadata (to avoid messing up colours)

jinjagit committed:
[TAN-4989] The basic idea, using exiftool

jinjagit committed:
[TAN-4989] Silence failing tests (for now)

jinjagit committed:
[TAN-4989] Add test + linting

jinjagit committed:
[TAN-4989] Add exiftool to production Dockerfile

jinjagit committed:
[TAN-4989] Improve test + more linting

jinjagit committed:
[TAN-4989] Update test of Orientation tag, and remove redundant permission of Rotation tag

jinjagit committed:
[TAN-4989] Remove unused constant

jinjagit commented on back/Dockerfile:
```diff
@@ -37,6 +37,8 @@ RUN apt-get update                                        \
    optipng                                                \
    pngquant                                               \
+   # Used to selectively strip EXIF metadata from images
+   libimage-exiftool-perl                                 \
```
I'm not sure how to test the production Dockerfile before deploying, so I'm just assuming this works (clearly it does in the `Dockerfile.development`)

jinjagit commented:
@kogre Tagging you just FYI, so you can check the changes if you wish.

jinjagit committed:
[TAN-4989] Improve comment

adessy commented on back/Dockerfile:
```diff
@@ -37,6 +37,8 @@ RUN apt-get update                                        \
    jpegoptim                                              \
    optipng                                                \
    pngquant                                               \
+   # Used to selectively strip EXIF metadata from images
```
Thanks for the comment :)

adessy commented on back/app/uploaders/base_image_uploader.rb:
```diff
@@ -83,25 +82,8 @@ def resize_to_fill_using_convert(img, target_width, target_height, gravity = 'Ce
-      img.strip
-      img
-    end
+    system("exiftool -all= -tagsFromFile @ -icc_profile -orientation -overwrite_original #{Shellwords.escape(@file.path)}")
```
Did you test it while using S3 as the storage backend?

Also, it would be great to check the output of the command and report it to Sentry if it fails. Otherwise, it might fail silently forever :)

jinjagit replied:
Not tested with s3. Good idea.

jinjagit replied:
Tested now (from dev env). Seems to work.

I used s3: true in env, restarted, uploaded idea image via UI... found image on s3 and downloaded it.
Then I used exiftool in terminal to check the tags for each image.
Original image has 137 EXIF tags, uploaded/downloaded image has 56 tags, and many of the tags removed are potential PII data.

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -3,6 +3,7 @@
 require 'carrierwave/test/matchers'
 require 'rails_helper'
 require 'mini_magick'
+require 'English'
```
What's that?

adessy replied:
Nevermind, I just checked it out. Learned something new :) Take a look at `Open3` to avoid relying on the global variables.

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,83 @@
+      # Check the command executed successfully
+      expect($CHILD_STATUS.success?).to be true
+
+      # Tags that are  technical/structural image metadata,
```
```suggestion
      # Tags that are technical/structural image metadata,
```

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,83 @@
+          tag.include?('Rendering') ||
+          tag.include?('Adaptation')
+        end
+      ).to be_empty
```
I don't quite understand what we're testing here. Isn't it sufficient to test `image.exif.keys`, or does that contain less information? Testing against the tool's output feels a bit brittle.

If we really need this test, I think it'd be helpful to include an example of the output.

jinjagit replied:
@adessy 

You mean the output of `exiftool_output = `exiftool #{Shellwords.escape(uploader.file.path)}``?

jinjagit replied:
I've added examples in comments + also improved the comments to explain why we use the tool in the test.

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,83 @@
+      ]
+      expect(image.exif.keys - allowed_exif_keys).to be_empty
+
+      exiftool_output = `exiftool #{Shellwords.escape(uploader.file.path)}`
```
`Open3.capture3` could be a better option here? See https://docs.ruby-lang.org/en/master/Open3.html#method-c-capture3.

adessy reviewed (commented):
Looks good to me! I’d just suggest improving the error reporting, and I also asked a few questions around testing.

jinjagit committed:
[TAN-4989] Add example of exiftool output in tests and improve test comments

jinjagit committed:
[TAN-4989] Use open3 for the command line operations in the test

jinjagit committed:
[TAN-4989] Use open3 in strip method + add error reporting & raising

jinjagit committed:
[TAN-4989] Remove non-open3 command call from test

jinjagit committed:
[TAN-4989] Linting

jinjagit committed:
[TAN-4989] Fix wrong ixture image file extensions detected by exiftool

jinjagit committed:
[TAN-4989] Fix tests that eference updated fixture image files

jinjagit commented:
@adessy I have addressed all of your suggestions. Thanks for them!

- We now have error reporting, and we raise an error if the `exiftool` command does not succeed, halting the execution of the image upload and ensuring a useful error is seen in the browser for the failed request.
- Now using `Open3.capture3` to run the `efitool` command instead of naive backticks approach.
- Test now includes examples of EXIF tags listed by `exiftool`, as well as improved comments.

I also did 3 more things:
- The new error raising highlighted that a number of fixture image files we use had the wrong extension (`.png`, when they are actually JPEGs). I changed the file extensions accordingly, and updated all the references to them in the code - mainly in seeding and tests. We (I) should remember this behaviour in cases where we need to re-upload existing images, as we may encounter this error for any existing image files with wrong extension types.
- I added a helper method to contain the call the `Open3.capture3`. This feels like maybe it should live in a service somewhere. WDYT? Maybe a new `ExternalCommandRunner` service?
- I created a separate [ticket](https://www.notion.so/govocal/Sentry-error-reporting-failing-in-dev-env-only-2349663b7b268053b92cf6c7b52c6c7f) to look into the failing Sentry errors we noticed. Would you mind taking a look and adding any info you think might be helpful? 🙏

adessy commented on back/app/uploaders/base_image_uploader.rb:
```diff
@@ -83,25 +84,72 @@ def resize_to_fill_using_convert(img, target_width, target_height, gravity = 'Ce
+        }
+      )
+      raise # Re-raise the original error
     end
```
I think this is going a step too far. I would just call `Open3.capture3` directly and, if an exception is raised, let it bubble up.

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,113 @@
+      args = [uploader.file.path]
+      stdout_str, stderr_str, status = Open3.capture3(command, *args)
+      expect(status.success?).to(be(true), "exiftool command failed: #{stderr_str}")
+      exiftool_output = stdout_str # Now, exiftool_output holds the stdout
```
```suggestion
      exiftool_output, stderr_str, status = Open3.capture3('exiftool', uploader.file.path)
      expect(status).to be_success, "exiftool command failed: #{stderr_str}"
```

I didn't know about custom messages. Just learned something new :)

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,113 @@
+          # or other non-harmful empty tags after stripping. Be flexible for these.
+          (tag.include?('Notes') && exiftool_output.include?("#{tag}:")) # Check if it's an empty "Notes" tag
+        end
+      ).to be_empty
```
I'm not a huge fan of this expectation. I find it hard to understand what it's testing, and it feels a bit brittle to me. But I suppose we don't expect to update ExifTool too frequently, do we?

jinjagit replied:
Yes, it is brittle, as will sometimes expose changes in the `exiftool` version we pick up.

We could pin the version of `exiftool` we use in the Dockerfile, or leave it as it is, and see how annoying it becomes.

@adessy WDYT?

adessy replied:
That's one option, but we are not doing it for any other deps (questionable, but saves us from having to actively monitor security upgrades for them).

I just found out that the `exiftool` has an option to output json, which would already make the test much less brittle.

jinjagit replied:
Ooh, OK, I'll take a look at that. Thanks.

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,113 @@
+
+      present_tags.reject! do |tag|
+        tag.empty? || tag == uploader.file.path.split('/').last
+      end
```
```suggestion
      filename = uploader.file.path.split('/').last

      present_tags = exiftool_output.lines.filter_map do |l|
        tag = l.split(':').first.strip
        tag unless tag.empty? || tag == filename
      end
```

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,113 @@
+          tag.include?('CMM') ||
+          tag.include?('Device') ||
+          tag.include?('Rendering') ||
+          tag.include?('Adaptation') ||
```
Nitpicky: If we cannot rely on equality here, shall we use a regex instead of all those `include?`.

jinjagit replied:
Now using an explicit list of full, expected tags, so no longer need to use `include?`

adessy commented on back/spec/uploaders/base_image_uploader_spec.rb:
```diff
@@ -28,7 +29,113 @@
+      # These are the tags that *should* be gone (privacy-sensitive, user-defined)
+      # after stripping, but are not in the `technical_tags` list.
+      # We allow ICC Profile and Orientation related tags to remain.
+      metadata_tags_to_check = present_tags - technical_tags
```
If those are the tags that should be gone, why do you remove extra tags in the expectation? I find it confusing.

jinjagit replied:
I've reorganized and simplified tests (again) to help improve readability.

adessy reviewed (approved):
Okay for me! I just left a few small comments. Also, while re-reading the `Open3.capture3` documentation, I came across this:

> Like Process.spawn, this method has potential security vulnerabilities if called with untrusted input; see Command Injection.

Might be good to give some extra thought to whether we're exposed to this risk. (In any case, it's probably no less secure than the backticks solution.)

jinjagit committed:
[TAN-4989] Convert seed image to JPEG (must all be same type)

jinjagit committed:
Merge branch 'master' into TAN-4989-selective-image-metadata-whitelist

jinjagit committed:
[TAN-4989] Improve test code readability

jinjagit committed:
[TAN-4989] Simplfy test code

jinjagit committed:
[TAN-4989] Remove extra error reporting when executing exiftool command

jinjagit committed:
[TAN-4989] Simplify and improve tests (simpler to understand/parse)

jinjagit committed:
[TAN-4989] Linting

jinjagit commented:
> Also, while re-reading the `Open3.capture3` documentation, I came across this:
> 
> > Like Process.spawn, this method has potential security vulnerabilities if called with untrusted input; see Command Injection.
> 
> Might be good to give some extra thought to whether we're exposed to this risk. (In any case, it's probably no less secure than the backticks solution.)

@adessy Here's a brief explanation of why `Open3.capture3(command, *args)`, which we use here, is safer against command injection:

When `Open3.capture3` (or `Process.spawn`) is called with multiple arguments (e.g. `command, arg1, arg2, ...`), it bypasses the shell. This means arguments are passed directly to the command's executable, treating any shell metacharacters within the arguments (like in `@file.path`) as literal strings rather than commands. This prevents malicious input from executing arbitrary code.

From https://docs.ruby-lang.org/en/3.3/Open3.html:

Example:
```Ruby
Open3.capture3('/usr/bin/date')
# => ["Thu Sep 28 05:03:51 PM CDT 2023\n", "", #<Process::Status: pid 2282300 exit 0>]
```
Ruby invokes the executable directly, with no shell and no shell expansion:
```Ruby
Open3.capture3('doesnt_exist') # Raises Errno::ENOENT
```
If one or more args is given, each is an argument or option to be passed to the executable:
```Ruby
Open3.capture3('echo', 'C #')
# => ["C #\n", "", #<Process::Status: pid 2282368 exit 0>]
Open3.capture3('echo', 'hello', 'world')
# => ["hello world\n", "", #<Process::Status: pid 2282372 exit 0>]
```

jinjagit committed:
[TAN-4989] Use exiftool's JSON output in the test

jinjagit committed:
[TAN-4989] Improve comment
